applications/GeoMechanicsApplication/custom_utilities/nodal_extrapolator.cpp

```diff
@@ -0,0 +1,111 @@
+                                                              GeometryData::IntegrationMethod IntegrationMethod) const
+{
+    CheckIfGeometryIsSupported(rGeometry);
+
+    auto          p_lower_order_geometry = CreateLowerOrderGeometry(rGeometry);
+    GeometryType& p_corner_geometry = p_lower_order_geometry ? *p_lower_order_geometry : rGeometry;
+
+    Matrix extrapolation_matrix =
+        CalculateExtrapolationMatrixForCornerNodes(rGeometry, IntegrationMethod, p_corner_geometry);
+
+    if (p_lower_order_geometry) {
+        AddRowsForMidsideNodes(rGeometry, extrapolation_matrix);
+    }
+
+    return extrapolation_matrix;
+}
+
+Matrix NodalExtrapolator::CalculateExtrapolationMatrixForCornerNodes(
+    const NodalExtrapolator::GeometryType& rGeometry,
+    GeometryData::IntegrationMethod&       IntegrationMethod,
+    const NodalExtrapolator::GeometryType& rCornerGeometry) const
+{
+    SizeType number_of_corner_nodes = rCornerGeometry.PointsNumber();
+
+    Matrix      quasi_mass_mat     = ZeroMatrix(number_of_corner_nodes, number_of_corner_nodes);
+    const auto& integration_points = rGeometry.IntegrationPoints(IntegrationMethod);
+    const auto  number_of_integration_points = integration_points.size();
+    Matrix      node_coefficient(number_of_corner_nodes, number_of_integration_points);
+    Vector      determinants_of_jacobian;
+    determinants_of_jacobian = rGeometry.DeterminantOfJacobian(determinants_of_jacobian, IntegrationMethod);
+
+    const Matrix& shape_functions_values_at_gauss_points =
+        rCornerGeometry.ShapeFunctionsValues(IntegrationMethod);
+
+    for (IndexType i_gauss_point = 0; i_gauss_point < number_of_integration_points; ++i_gauss_point) {
+        const Vector N = row(shape_functions_values_at_gauss_points, i_gauss_point);
+        quasi_mass_mat += outer_prod(N, N) * determinants_of_jacobian[i_gauss_point] *
+                          integration_points[i_gauss_point].Weight();
+        column(node_coefficient, i_gauss_point) =
+            N * determinants_of_jacobian[i_gauss_point] * integration_points[i_gauss_point].Weight();
+    }
+
+    double metric_determinant;
+    Matrix quasi_mass_mat_inverse;
+    MathUtils<double>::InvertMatrix(quasi_mass_mat, quasi_mass_mat_inverse, metric_determinant, -1.);
+
+    auto extrapolation_matrix = Matrix{prod(quasi_mass_mat_inverse, node_coefficient)};
+    return extrapolation_matrix;
+}
+
+void NodalExtrapolator::AddRowsForMidsideNodes(const NodalExtrapolator::GeometryType& rGeometry,
+                                               Matrix& extrapolation_matrix) const
+{
+    const size_t n_filled = extrapolation_matrix.size2();
+    extrapolation_matrix.resize(rGeometry.PointsNumber(), extrapolation_matrix.size2());
+    for (size_t i_row = 0; i_row < rGeometry.PointsNumber() - n_filled; ++i_row) {
+        row(extrapolation_matrix, n_filled + i_row) =
+            0.5 * (row(extrapolation_matrix, i_row) + row(extrapolation_matrix, (i_row + 1) % n_filled));
```

Comment:
This statement becomes dependent on element type and which midside node when we implement solid elements.